src/inmanta/server/services/orchestrationservice.py

```diff
@@ -63,277 +63,299 @@
 # Kill switch for cleanup, for use when working with historical data
 
 
-class ResourceWithResourceSet:
-    def __init__(
-        self,
-        resource: ResourceMinimal,
-        resource_set: Optional[str],
-        resource_state: const.ResourceState,
-    ) -> None:
-        self.resource = resource
-        self.resource_set = resource_set
-        self.resource_state = resource_state
+class CrossResourceSetDependencyError(Exception):
+    def __init__(self, resource_id1: ResourceIdStr, resource_id2: ResourceIdStr) -> None:
+        """
+        Raised when a cross-resource set dependency was detected between the resource with id
+        resource_id1 and resource_id2.
+        """
+        self.resource_id1 = resource_id1
+        self.resource_id2 = resource_id2
+        super().__init__(self.get_error_message())
+
+    def get_error_message(self) -> str:
+        return (
+            f"A dependency exists between resources {self.resource_id1} and {self.resource_id2}, but they belong to"
+            f" different resource sets."
+        )
+
 
-    def get_resource_id_str(self) -> ResourceIdStr:
-        return self.resource.get_resource_id_str()
+class ResourceSetValidator:
+    def __init__(self, resources: abc.Set[data.Resource]) -> None:
+        self.resources = resources
+        self.rid_to_resource_set = {res.resource_id: res.resource_set for res in self.resources}
 
-    def get_resource_state_for_new_resource(self) -> Literal[ResourceState.undefined, ResourceState.available]:
+    def _is_cross_resource_set_dependency(self, res: data.Resource, rid_dependency: ResourceIdStr) -> bool:
         """
-        Return the resource state as expected by the `resource_state` argument of the `OrchestrationService._put_version()`
-        method. This method should set the resource state of a resource to undefined when it directly depends on an unknown or
-        available when it doesn't.
+        Return True iff the dependency between resource res and the resource with id rid_dependency is a cross-resource set
+        dependency.
         """
-        if self.resource_state is ResourceState.undefined:
-            return ResourceState.undefined
-        else:
-            return ResourceState.available
-
-    def is_shared_resource(self) -> bool:
-        return self.resource_set is None
+        if res.resource_set is None:
+            # Resource is in shared resource set.
+            return False
```

Comment:
# pass I
A dependency starting in the shared resource set should remain in the shared resource set.

So I think this would allow the shared set to refer back to others,...

# pass II
Looking deeper into the code, it seems that this IS supported, which is pretty cool, but it will take some docs